packages/react-admin-core/src/stack/Switch.tsx

```diff
@@ -33,104 +27,97 @@ export interface IStackSwitchApi {
-                                        return (
-                                            <StackBreadcrumb
-                                                url={props.match.url}
-                                                title={this.state.pageBreadcrumbTitle[page.props.name] || page.props.title || page.props.name}
-                                            >
-                                                {ret}
-                                            </StackBreadcrumb>
-                                        );
-                                    }
+    if (!match) return null;
+
+    return React.Children.map(props.children, (page: any) => {
+        const path = isInitialPage(page.props.name) ? `${match.url}` : `${match.url}/:id/${page.props.name}`;
+        return (
+            <Route path={path} exact={isInitialPage(page.props.name)}>
+                {(routeProps: RouteComponentProps<IRouteParams>) => {
+                    if (!routeProps.match) return null;
+                    activePage = page.props.name;
+                    const ret = (
+                        <StackSwitchMeta
+                            id={id}
+                            activePage={page.props.name}
+                            isInitialPageActive={isInitialPage(page.props.name)}
+                        >
+                            <StackSwitchApiContext.Provider
+                                value={{
+                                    activatePage,
+                                    id,
+                                    updatePageBreadcrumbTitle,
                                 }}
-                            </Route>
+                            >
+                                {typeof page.props.children === "function"
+                                    ? page.props.children(routeProps.match.params.id)
+                                    : page.props.children}
+                            </StackSwitchApiContext.Provider>
+                        </StackSwitchMeta>
+                    );
+                    if (isInitialPage(page.props.name)) {
+                        return ret;
+                    } else {
+                        return (
+                            <StackBreadcrumb
+                                url={routeProps.match.url}
+                                title={pageBreadcrumbTitle[page.props.name] || page.props.title || page.props.name}
+                            >
+                                {ret}
+                            </StackBreadcrumb>
                         );
-                    });
+                    }
                 }}
             </Route>
         );
-    }
-    public activatePage = (pageName: string, payload: string, subUrl?: string) => {
-        if (this.isInitialPage(pageName)) {
-            this.history.push(this.match.url);
+    });
+
+
+
+    function activatePage(pageName: string, payload: string, subUrl?: string) {
```

Comment:
Private functions in einer React-Componente sollten generell vor dem return-statement definiert sein.
Warum ?
Wenn man die Funktionen refactoren möchte als useCallback() dann müssen sie über dem return sein wegen "rule-of-hooks".
Wir sollten generell viel mehr useCallback verwenden, aber das diskutieren wir an anderer Stelle.

Reply:
useCallback bin ich auch dafür 👍